### PR TITLE
feat: implement dev server as plugin

### PR DESCRIPTION
**Pull request by rishabh3112**
- [x] This is a **bugfix**
- [x] This is a **feature**
- [x] This is a **code refactor**
- [ ] This is a **test update**
- [x] This is a **docs update**
- [x] This is a **metadata update**

### Motivation / Use-Case
The Dev Server should be implemented as a plugin for long run. This would simplify implementation of `webpack serve` as will give user more flexibility in multicompiler case.

The current implementation code present in `DevServerPlugin.js` implements few parts but still relies on `Server.js` to be applied.

I am currently implementing it separately at `WebpackDevServerPlugin.js`  to simplify testing. Later on we will export only this plugin instead on `Serve.js`.

### Breaking Changes
Yes, it would be a breaking change.

### Additional Info
@alexander-akait suggested this feature. Rest, we can chat on this PR.

### CODE REVIEW

**Review by alexander-akait — COMMENTED** [review]
Logic should be in `apply` function here https://github.com/webpack/webpack-dev-server/blob/master/lib/Server.js, no need to duplicate this, all method already exists in server

File: lib/Server.js [hunk 1]
```diff
@@ -39,18 +39,21 @@ if (!process.env.WEBPACK_SERVE) {
 }
 
 class Server {
-  constructor(compiler, options = {}) {
```
Comment (alexander-akait) [inline lib/Server.js:42]: We need keep compatibility here, just ignore compiler from constructor for apply

Comment (alexander-akait) [reply lib/Server.js:42]: Ideally after stable release v4, we will release 4.1/4.2 with plugin support, and for v5 we drop compiler from `constructor`

Comment (alexander-akait) [reply lib/Server.js:42]: Sounds good, we need do it before stable release, let's do it separately and fix in webpack-cli too, maybe adding hidden static option `_newConstructor`

Comment (rishabh3112) [reply lib/Server.js:42]: On it 👍🏽

**Comment by rishabh3112** [comment]
## Status
Waiting merge of https://github.com/webpack/webpack-dev-server/pull/3273

File: lib/Server.js [hunk 2]
```diff
@@ -39,24 +39,34 @@ if (!process.env.WEBPACK_SERVE) {
...
+      this.apply(options);
+    }
+  }
+
+  apply(compiler) {
+    this.compiler = compiler;
+    this.logger = this.compiler.getInfrastructureLogger(this.name);
+
     normalizeOptions(this.compiler, this.options);
     this.applyDevServerPlugin();
```
Comment (alexander-akait) [inline lib/Server.js:71]: Ideally I think we need inline this helper here, let's do it after https://github.com/webpack/webpack-dev-server/pull/3260/ and some refactoring

Anyway good job and thanks

Comment (rishabh3112) [reply lib/Server.js:71]: Yeah removing `DevServerPlugin` and perform ops directly on child compilers is next step to do in my list.

File: lib/Server.js [hunk 3]
```diff
@@ -74,14 +84,218 @@ class Server {
...
+        portString = `&port=${options.webSocketServer.options.port}`;
+      } else {
+        portString = `&port=${options.port}`;
+      }
+
+      /** @type {string} */
+      let pathString = '';
+
+      // Only add the path if it is not default
+      if (options.client && options.client.path && options.client.path) {
```
Comment (snitin315) [inline lib/Server.js:131]:
```suggestion
      if (options.client && options.client.path) {
```

**Comment by rishabh3112** [comment]
@alexander-akait what should I do next in this PR. I am done on my end 😅

**Review by alexander-akait — COMMENTED** [review]
Can we avoid a lot of change, maybe better create a new PR

**Comment by rishabh3112** [comment]
> Can we avoid a lot of change, maybe better create a new PR
Yup, will create a new PR for this now.

**Pull request closed without merging**